restapi/src/main/java/io/stargate/web/docsapi/service/query/DocumentSearchService.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.stargate.web.docsapi.service.query;
+
+import com.bpodgursky.jbool_expressions.Expression;
+import io.reactivex.rxjava3.core.Flowable;
+import io.reactivex.rxjava3.core.Single;
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.query.BoundQuery;
+import io.stargate.db.query.Query;
+import io.stargate.db.query.builder.BuiltQuery;
+import io.stargate.web.docsapi.dao.Paginator;
+import io.stargate.web.docsapi.service.DocsApiConfiguration;
+import io.stargate.web.docsapi.service.ExecutionContext;
+import io.stargate.web.docsapi.service.QueryExecutor;
+import io.stargate.web.docsapi.service.RawDocument;
+import io.stargate.web.docsapi.service.query.search.db.impl.PopulateSearchQueryBuilder;
+import io.stargate.web.docsapi.service.query.search.resolver.BaseResolver;
+import io.stargate.web.docsapi.service.query.search.resolver.DocumentsResolver;
+import io.stargate.web.docsapi.service.query.search.weigth.ExpressionWeightResolver;
+import io.stargate.web.rx.RxUtils;
+import java.util.Collection;
+import javax.inject.Inject;
+import org.apache.commons.lang3.tuple.Pair;
+
+public class DocumentSearchService {
+
+  @Inject private ExpressionWeightResolver<FilterExpression> weightResolver;
+  @Inject private DocsApiConfiguration configuration;
+
+  public Flowable<RawDocument> searchDocuments(
+      QueryExecutor queryExecutor,
+      String keyspace,
+      String collection,
+      Expression<FilterExpression> expression,
+      Collection<String> fields,
+      Paginator paginator,
+      ExecutionContext context) {
+
+    // resolve the expression
+    DocumentsResolver documentsResolver = BaseResolver.resolve(expression, weightResolver, context);
+
+    // load the candidates
+    Flowable<RawDocument> candidates =
+        documentsResolver
+            .getDocuments(queryExecutor, configuration, keyspace, collection, paginator)
+
+            // limit to requested page size only to stop fetching extra docs
+            .take(paginator.docPageSize);
+
+    // then populate
+    return populateCandidates(
+        candidates, queryExecutor, keyspace, collection, nestedPopulate(context));
```

Comment:
As I mentioned in my other comment, if `where` is empty, we are going to load each doc twice.

I'd propose to handle the case where no predicates are provided outside of `BaseResolver.resolve` because is that case we do not really need to "search and then load", we load all docs from one query. WDYT?

Reply:
Yea this is awesome catch, failed into the trap of the generalizing the solution..

Reply:
Done, pushed to the docs service as this is not a resolver anymore imo..